dev-libs/wlc: add ebuilds to repo.

tajjada opened the pull request:
Package-Manager: portage-2.2.28

I would like to add this new package to gentoo.

BTW: I also tried to add multilib support to these ebuilds, but could not, because the libinput ebuild in gentoo does not support multilib, and it is a dependency.

tajjada commented:
I have no idea how these QA bot checks work, but aren't all the above messages from other packages, and not relevant to my ebuild/PR?

Do I need to worry about them? Did I do anything wrong?

rindeal commented:
> Do I need to worry about them? Did I do anything wrong?

Nope, it's just spam

idella commented:
correct

idella commented:
So these are brand new ebuilds.  The commit msg ought reflect that it is one package, offering or including a live and a versioned ebuild, and cite the version (0.0.2)  This will take some testing

Hummer12007 commented:
Isn't wayland USE flag necessary for media-libs/mesa? (it doesn't build the wayland egl platform otherwise)

idella commented:
My initial attempt to emerge the deps resulted in a reject by portage:

media-libs/mesa[gbm,gles2,egl]

```bash
USE="gbm gles2 egl" emerge -1 media-libs/mesa -pv

  (dev-qt/qtgui-5.6.0-r1:5/5.6::gentoo, ebuild scheduled for merge) conflicts with
    ~dev-qt/qtgui-5.6.0[-gles2,-gtkstyle,png,xcb] required by (dev-qt/qtwidgets-5.6.0:5/5.6::gentoo, installed)
```

This combination alone will cause difficulties. Someone else will need to pursue with this at this point.  I will not unmerge half my installed packages for the sake of one runtest

tajjada commented:
@Hummer12007 

Added the wayland USE flag dependency, as you suggested.

I also added more USE flags and improved src_configure.

Thanks for the suggestions.

monsieurp commented:
Hi. Did you file a bug for this new package?

tajjada commented:
@monsieurp 

... uhm ... no ... was I supposed to?

Basically, I wanted to add an ebuild for this library to gentoo (and this is my first time trying to get an ebuild into official repos), so I read the wiki pages about Proxy Maintainers and Gentoo Github, followed what was written there, and made my pull request. Is filing a bug request required?

monsieurp commented:
Yes please. Just to make sure you are reachable and your email address in our bug tracking system i.e. https://bugs.gentoo.org.

tajjada commented:
Done: https://bugs.gentoo.org/show_bug.cgi?id=582386

mjeveritt commented:
@tajjada you may find the multilib eclass helpful

idella commented:
tajjada yes it is helpful in the gentoo distro overall to mirror this pull req with a bug in bugzilla. Bugzilla came this, and github extends to service of a path to add a package

idella commented:
Oh under certain conditions you can exempt the one lib / package form complying with mutlilib building.
It was pointed out by a dev in g-dev. If you cannot figure it get back to me.

tajjada commented:
@mjeveritt @idella 

As I mentioned in the first comment on this PR, I *did* attempt to add multilib support to this ebuild. However, it didn't work out, because wlc depends on libinput, and the libinput package in gentoo does not currently have multilib.

I didn't want to modify other people's ebuilds (to add multilib support to libinput), so I submitted this PR without multilib.

mjeveritt commented:
@tajjada : sorry, my bad. You -could- file a bug against libinput if you felt it necessary, but looks like you've got things under control :) hopefully a merge will be forthcoming soon.

idella commented:
In my system this 
media-libs/mesa[wayland,gbm,gles2,egl]
is simply not viable

from gst-plugins-bad-1.6.3, a dep pulled in, it sets
egl? ( !gles2 )

making an a attempt to
# USE="wayland gbm gles2 egl" emerge media-libs/mesa
unviable.  
While it may be viable on other systems, the combo need suite an average user's system.
These use flags need a revision

tajjada commented:
@idella 

wlc's [README](https://github.com/Cloudef/wlc/blob/master/README.rst) clearly specifies that it wants both GLES2 and EGL installed as dependencies. This is why I listed both of those USE flags.

The combination of USE flags seems fine on my system (but I don't use gstreamer, which seems to be the problematic one, from your comment). That said, I don't understand why you consider your system to be representative of "an average user's system", as every gentoo user has a fairly different set of USE flags (kind of the whole point of using gentoo :) ), but I will look into it. I will experiment with trying to build wlc with mesa having a different combination of USE flags, but I am not sure if it will work. If wlc builds with either -gles2 or -egl, I will adjust the ebuild accordingly.

tajjada commented:
OK, here is what I found:

mesa has: REQUIRED_USE="gles2? ( egl ) wayland? ( egl )", so it clearly *cannot* be installed with USE="gles2 -egl", so that is not an option.

USE="gles2" on mesa is required by lots and lots of other wayland-related packages, too (weston, libsdl2 with wayland support, etc). Wayland users will need to have USE="gles2 wayland egl" anyway, because lots wayland-related things tend to require that. Same with wlc. There is nothing wrong with my ebuild, and the USE flag combination you pointed out as being "unviable" is actually very typical for a Wayland system.

I guess that gst-plugins-bad-1.6.3 just doesn't play along with gles2, and that is gstreamer's problem. I don't see how that affects wlc or my ebuild, or anything else that requires the same combination of USE flags (weston and libsdl2[wayland] are two examples I have on my system).

mjeveritt commented:
I would agree the problem is with gst-plugins-bad .. and its for this reasoned the package is named as such.

Hummer12007 commented:
@tajjada the issue is in fact not direcrly in the ebuild, but in manually installing the dependencies (via the USE="..." emerge media-libs/mesa command), as opposed to pulling them in via dependency resolution

idella commented:
So it could be a technical inconvenience. I shall retry tomorrow.
Installing the deps manually separately is something I do all the time.  This would be the first time I have encountered such a technical tripup. But stranger things have happened.

idella commented:
yes, on persevering, running the ebuild from emerge as you 'said' is viable. It does require settings customised entries into /etc/portage/package.use/package.use, which is not unviable.  Yet to fully runtest at this stage.

idella commented:
with mesa emerged with those USE flags set this builds for me. At this point I believe you can possibly add multilib support and exclude the need in  libinput .

As for the build failure, although your system builds it, the design behind ebuilding is to address any potential system. Despite that it builds without -fpic, it needs it in the build on mine, therefore it will need it on other potential users' systems.  CAn you edit the ebuild to append -fpic to the required build FLAGS and re-push. This ebuild is looking good but being a highly specialised package, will require further input form those similarly specialised. This is normal

tajjada commented:
@idella 

Regarding multilib support, why would I add it, when it breaks the build? When I tried building wlc with multilib, it failed with linker errors due to the linker not being able to find 32-bit libinput.so. What do you mean by "exclude the need in libinput"?

Also, what did you mean about `-fPIC` exactly? I just ran `ebuild wlc-0.0.2.ebuild compile`, and looked at the compiler flags used during the compilation process. The build system seems to be adding `-fPIC` correctly (everything is built with `-fPIC`, and I did not add it manually anywhere (make .conf / package.env / the ebuild), so it must have been added by the build system (as it should be). Did you mean something else? I don't understand what you suggested needed to be changed.

Polynomial-C commented on dev-libs/wlc/wlc-0.0.2.ebuild:
```diff
@@ -0,0 +1,50 @@
+# Distributed under the terms of the GNU General Public License v2
+# $Id$
+
+EAPI=5
```
Why EAPI-5 and not EAPI-6?

idella commented:
@tajjada forget -fpic. It was a glitch in my setup. It builds in my system now without concern over -fpic, The suggestion re -fpic came form an error message from the build system which is simply misleading.

If you want to add multilib support it is still possible. However adding it via comments in this page makes it very awkward. Do you want to have multilib support added or not.  The setup in gentoo to achieve it is not something I can type out in a few lines.  If you do, I suggest you join the  IRC channel cited in 

https://wiki.gentoo.org/wiki/Project:Proxy_Maintainers

tajjada commented:
@idella 

Regarding multilib support:

I want to make my ebuilds as good as possible. Multilib is a good thing, so I am certainly interested in adding it. I just thought that it wasn't possible, due to dependencies (libinput) missing multilib support. Because of this, I imagined the best thing to do would be to leave the ebuild without multilib support for now, file a bug against libinput, and then update the ebuild to add multilib support in the future, when libinput has it.

However, you clearly know things that I don't. I joined both #gentoo-dev-help and #gentoo-proxy-maint IRC channels. I am interested to find out what you have in mind.

axs-gentoo commented:
@tajjada multilib support should only be added if there is an actual use-case or need for it.  Do you have or know of packages that will need 32bit versions of the libraries installed on a 64bit system?  If not, then there's no point even considering multilib support at this time, imo.

When multilib support is required, then it can be done on a lib-by-lib basis -- that is, if only one component needs an alt-abi xinput but others are fine, then its fine to only build the alt-abi components that ave supported by alt-abi deps.  However, again, this all depends on what rdeps might need.  Also note that partial builds for alt-abi's can get messy (one more reason why it likely shouldn't be attempted at this point).

Final (non-multilib) note -- it does make sense to make this package EAPI6, which in this case I believe means converting those cmake-utils_use-* calls to whatever is EAPI6 compliant.

tajjada commented:
@axs-gentoo 

While I agree with you that multilib isn't *necessary*, gentoo is all about customisability, and different users use different things. It is nice for ebuilds to provide options for users; maybe someone at some point might want to have multilib wlc, even if nothing else in portage actually depends on it. IMO it would be nice if all libs in gentoo had multilib support.

That said, I agree that *at this time* it isn't necessary, and the lack of multilib shouldn't prevent my ebuild from being merged into portage. It would still be a nice extra feature to consider adding at some point in the future.

I am also interested in learning new things about writing gentoo ebuilds, so I wanted to find out what @idella had in mind about multilib.

Regarding EAPI=6, I will look into it.

tajjada commented:
@axs-gentoo 

EAPI=6 is done.

idella commented:
firstly axs-gentoo thanks alot for your input here.
To mu understanding it is feasible to add multilib support while keeping to the one set _ABI of the yet unsupported lib "libinput".  What I need form you is a decision as to whether you wish to pursue having it added on addition to the tree.  For this you have to start inheriting the required eclass of multilib (there are a few), setting MULTILIB_USEDEP to the deps in python eclass style, and work up to a viable state.  You could also simply submit a bug to the maintainers of libinput to support multilib, and take that path.
Use of irc and live interaction cuts down the  time taken to address these by large margins. It seems you arrived shortly after I hibernated computer. Yes my tz is atypical.

idella commented:
frankly considering this will be a new addition to the tree, these various commits ought be squashed into 1 or 2.  The initial commit of an ebuild is its final state and doesn't require the history of the changes during testing and feedback to get it there/\.

tajjada commented:
Done. Squashed everything into one commit.

idella commented:
This has now had extensive review and discussion both in this pull req and IRC in the channel.
Adding
